pkg/cloudprovider/provider/packet/provider.go

```diff
@@ -0,0 +1,420 @@
+		}
+	}
+
+	tag := generateTag(string(machine.UID))
+	client := getClient(c.APIKey)
+	device, err := getDeviceByTag(client, c.ProjectID, tag)
+	if err != nil {
+		return err
+	}
+
+	// append to our tags
+	tags := append(device.Tags, tag)
+
+	glog.Infof("Setting UID label for machine %s", machine.Name)
+	dur := &packngo.DeviceUpdateRequest{
+		Tags: &tags,
+	}
+	_, response, err := client.Devices.Update(device.ID, dur)
+	if err != nil {
+		return packetErrorToTerminalError(err, response, "failed to update UID label")
+	}
+	glog.Infof("Successfully set UID label for machine %s", machine.Name)
+
+	return nil
+}
+
+func (p *provider) GetCloudConfig(spec v1alpha1.MachineSpec) (config string, name string, err error) {
+	return "", "", nil
+}
+
+func (p *provider) MachineMetricsLabels(machine *v1alpha1.Machine) (map[string]string, error) {
+	labels := make(map[string]string)
+
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err == nil {
+		labels["size"] = c.InstanceType
+		labels["facilities"] = strings.Join(c.Facilities, ",")
+	}
+
+	return labels, err
+}
+
+func (p *provider) SetMetricsForMachines(machines v1alpha1.MachineList) error {
+	return nil
+}
+
+type packetDevice struct {
+	device *packngo.Device
+}
+
+func (s *packetDevice) Name() string {
+	return s.device.Hostname
+}
+
+func (s *packetDevice) ID() string {
+	return s.device.ID
+}
+
+func (s *packetDevice) Addresses() []string {
+	// returns addresses in CIDR format
+	var addresses []string
+	for _, ip := range s.device.Network {
+		addresses = append(addresses, fmt.Sprintf("%s/%d", ip.Address, ip.CIDR))
+	}
+
+	return addresses
+}
+
+func (s *packetDevice) Status() instance.Status {
+	switch s.device.State {
+	case "provisioning":
+		return instance.StatusCreating
+	case "active":
+		return instance.StatusRunning
+	default:
+		return instance.StatusUnknown
+	}
+}
+
+/******
+CONVENIENCE INTERNAL FUNCTIONS
+******/
+
+func getDeviceByTag(client *packngo.Client, projectID, tag string) (*packngo.Device, error) {
+	devices, response, err := client.Devices.List(projectID, nil)
+	if err != nil {
+		return nil, packetErrorToTerminalError(err, response, "failed to list devices")
+	}
+
+	for _, device := range devices {
+		if itemInList(device.Tags, tag) {
+			return &device, nil
+		}
+	}
+	return nil, nil
+}
+
+// given a defined Kubermatic constant for an operating system, return the canonical slug for Packet
+func getNameForOS(os providerconfig.OperatingSystem) (string, error) {
+	switch os {
+	case providerconfig.OperatingSystemUbuntu:
+		return "ubuntu_18_04", nil
+	case providerconfig.OperatingSystemCentOS:
+		return "centos_7", nil
+	case providerconfig.OperatingSystemCoreos:
+		return "coreos_stable", nil
+	}
+	return "", providerconfig.ErrOSNotSupported
+}
+
+func getClient(apiKey string) *packngo.Client {
+	return packngo.NewClientWithAuth("kubermatic", apiKey, nil)
+}
+
+func generateTag(ID string) string {
+	return fmt.Sprintf("gardener:%s:%s", machineUIDTag, ID)
```

Comment:
That does not feel right ;)

Reply:
I am more than happy to change it to something else. Suggestion?

Reply:
Change the `gardener` for `kubermatic` or just `machine-controller`?

Reply:
Doh! I copied/pasted a log message from somewhere else! Definite facepalm!